primitives/keystore/src/lib.rs

```diff
@@ -114,8 +145,8 @@ pub trait CryptoStore: Send + Sync {
 
 	/// Checks if the private keys for the given public key and key type combinations exist.
 	///
-	/// Returns `true` iff all private keys could be found.
-	async fn has_keys(&self, public_keys: &[(Vec<u8>, KeyTypeId)]) -> bool;
+	/// Returns the indices of the matching keys in `public_keys`.
```

Comment:
If you look two functions up, there is already a function called `supported_keys`. TBH  I find the (overall) API of Keystore very confusing now - it has completely different patterns and seemingly overlapping funcitonality.

Why can't we extend `supported_keys` or use a similar pattern to return an intersection?
Alternatively let's unify these functions and return `HasKeys` from `supported_keys` as well, asl would be worth emphasizing the differences.

Simply checking if all keys are found is just comparing the length of both collections and getting an empty collection handles `None`.

Reply:
Also `has_*` and `is_*` functions kind of indicate that `bool` is going to be returned, so returning a custom enum is quite confusing, I'd suggest changing the name a bit.

Reply:
I also don't like returning indices of the input Vec and agree that there's a overlap in functionality and an inconsistency in usage-patterns. 

Also in all places I adapted code to the new API, I had to first collect the keys into a `Vec<(Key, Id)>` to use the api. So if the id is the same in all cases, doesn't it make sense to just pass it as it's own parameter?

I also think the name `has_keys` implies a boolean context. So because of that and to get this thing landed as discussed [here](https://github.com/paritytech/grandpa-bridge-gadget/pull/148#issuecomment-820341426) I propose:

* Restoring (and maybe deprecating) the old `has_keys` function.
* Renaming the `HasKeys` enum to `FindKeys`, but containing the keys instead of indices.
* Renaming the current `has_keys` function to `find_keys` and using this signature: 
    ```rust
     fn find_keys(&self, id: KeyTypeId, public_keys: &[Vec<u8>]) -> FindKeys
   ```
Alternatively forget about all that, restore the old `has_keys` signature and use `supported_keys` with external checks for *all* and *empty*


Disclaimer: I'm new to substrate, so just my 2ct and humble opinion from what I know.

Reply:
Thanks for your contribution and the proposal. Sorry for suboptimal contributor experience you might have :) The issues with `bridge-gadget` are something new and we have to figure out ourselves how to best deal with them. The fact that it's an external contribution might complicate the situation a bit as well. Also the keystore API seems a bit messy already, hence no single vision on our side here.

Regarding your points:

> Restoring (and maybe deprecating) the old has_keys function.

Yes, IMHO that's a good idea to both solve the dependency issue and discuss the best replacement API.

> Renaming the HasKeys enum to FindKeys, but containing the keys instead of indices.

Returning keys directly instead of indices is definitely a way to go (currently you have to clone the initial vector of keys before passing it, to later retrieve what indices point to). I'm not fully convinced on having the enum though, cause it creates a bunch of implicit assumptions which might lead to a bug, examples:
1. Is `HasKeys::None == HasKeys::Found(vec![])`?
2. What if an empty vec was passed as an argument? Should it return `None`, `FoundAll()` or `Found(vec![])`?
3. Why distinguish between `Found` and `FoundAll`? Seems that `found_all` is only really useful for tests.

> Renaming the current has_keys function to find_keys and using this signature...

I like that idea as well, but I feel it now becomes dangerously similar with `supported_keys` method, hence I'd investigate if it's possible to merge these two as well. Perhaps it's possible to use `supported_keys` in place of `has_keys` already? If not then I'd create a method `supported_raw_keys(id: KeyTypeId, Vec<Vec<u8>>)` to indicate that these two methods are basically identical, just one works with `CryptoTypePublicPair` and second with raw (non-app-crypto) keys.
As mentioned earlier the `found_any/found_all/into_found` semantics are imho easily replaceble:
```rust
// found_any
if (!keystore.supported_keys(id, keys).is_empty()) {}

// found_all -- note it's only used for tests
let len = keys.len();
if (keystore.supported_keys(id, keys).len() == len) {}

// found none:
if (keystore.supported_keys(id, keys).is_empty()) {}

// into_found
for key in keystore.supported_keys(id, keys) {}
```

Reply:
So, it seems reasonable to restore and deprecate the old `has_keys` to make it possible to land this.

I'll try to implement the `has_keys`usages to use `supported_keys` and after this PR was merged, we can remove `has_keys` and change it's usages in a companion polkadot and BEEFY PR as well.

Before I put more work into this, I'd like to hear @bkchr 's opinion, as he was involved in the initial issue.

Reply:
I don't understand why we have this `supported_keys` method at all. The interface also looks weird.

However, it sounds good to merge both methods into one.

Regarding the return value, I wanted to have `indices` for a reason. In most places where we use this, we want to have the index. The point is that we fetch some runtime authorities, check which key we got and propagate the authority as its index.

Maybe the method could just take a slice, than we don't need to clone the keys.

Reply:
Ok, to get this landed I'll start another try. 

* I'll deprecate `has_keys` and replace all usages with equivalent `supported_keys` code.
* I'll create polkadot and beefy companions with the same changes
* We merge and then I'll change `supported_keys` to a better API which also return indicies in a followup PR.

I'm currently at the first bullet and got a problem [here](https://github.com/paritytech/substrate/blob/master/client/rpc/src/author/mod.rs#L131). When I convert the session keys to `CryptoTypePublicPairs`, can I just use any `CryptoTypeId` like e.g. `sp_core::ed25519::CRYPTO_ID`?

And is it correct that the `KeyTypeIds` of the session keys are all the same, so I can just take the first one?